### ISSUE TITLE
Storing thumbnails in memory is not friendly to memory constrained devices

### ISSUE THREAD

**Issue by lukaszsobala** [nomacs/nomacs#857]
**Describe the bug**
Nomacs (compiled on debian bullseye-based armv7l system) uses more and more memory until crashing with the following error (dmesg):
```
[24844.846367] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/,task=nomacs,pid=6546,uid=0
[24844.846470] Out of memory: Killed process 6546 (nomacs) total-vm:597892kB, anon-rss:0kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:522kB oom_score_adj:0
[24845.648599] oom_reaper: reaped process 6546 (nomacs), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
```

**To Reproduce**
1. Open nomacs (command line or not, doesn't matter) compiled on an armhf system (32-bit arm) 
2. Open an image in a folder with many images
3. Scroll for a long time or let the slideshow play for some time
4. Observe nomacs using more and more memory until crashing

**Expected behavior**
Nomacs' memory consumption should not rise indefinitely but until the set cache and history is full. 

**Screenshots**
Can add later but it's easy enough to reproduce

**Desktop:**
 - OS: DietPi (debian bullseye) 
 - Version 8.4

**Additional context**
I wanted to use nomacs instead of feh for a photo frame, as it has touch controls. It has a small enough memory footprint at the beginning, but no matter how large or how small I set the cache size to (even 0), nomacs seems to keep all the previously shown images in memory, eventually filling up all of the RAM and swap and then, crashing. The pi zero 2 w, which I use, should be able to run nomacs without a problem with a reasonable amount of cache (100 MB, for example). 

Cheers!

**Comment by lukaszsobala** [comment]
Is this repo alive?

**Comment by lukaszsobala** [comment]
Very disappointing.

**Comment by novomesk** [comment]
Dzień dobry,

I am sorry, but the nomacs project needs volunteers and developers. If you know someone, you may suggest them to join the project, even if they have little experiences now. This project offers many areas where people can learn.

I don't have 32-bit arm system, I can't use the steps to reproduce.

If you have time, create a debug build of nomacs and try to use **valgrind** to find potential sources of memory leaks.

https://www.cprogramming.com/debugging/valgrind.html

https://valgrind.org/docs/manual/quick-start.html

I don't guarantee it will find the source of the problem, but that is what I would try in your place.

**Comment by lukaszsobala** [comment]
Hi @novomesk,

thanks for responding!

I have very little knowledge of C programming but will try to create a debug build and use this valgrind tool. Maybe it's a simple fix. I hope some developers will resume this project at some point. I think this also happens on other architectures though, but I will have to test it as well.

Cheers,
Lukasz

**Comment by leejuyuu** [comment]
@lukaszsobala How large is your system memory?

**Comment by lukaszsobala** [comment]
The memory on raspberry pi zero 2w is 512 MB. But I do not see how this is relevant, as it will fill up any amount of memory.

**Comment by leejuyuu** [comment]
Try setting `maxImagesCached` in `[ResourceSettings]` to a small number like 1. The default is 5. The current implementation does not always respect the cache memory limit.

It's relevant because we test with PCs that have GBs of RAM, which makes us unable to reproduce OOM kill. Knowing more about your constraints helps us determine whether and how we can fix this.

**Comment by leejuyuu** [comment]
If possible, please also test setting the cache memory to 0. This will completely disable caching.

**Comment by scrubbbbs** [comment]
> which makes us unable to reproduce OOM kill

You can use the `ulimit` feature in bash to do this.

**Comment by lukaszsobala** [comment]
> Try setting `maxImagesCached` in `[ResourceSettings]` to a small number like 1. The default is 5. The current implementation does not always respect the cache memory limit.
> 
> It's relevant because we test with PCs that have GBs of RAM, which makes us unable to reproduce OOM kill. Knowing more about your constraints helps us determine whether and how we can fix this.

I don't think so, because you can easily monitor memory use and see if it goes above a certain value that you set. As I wrote in the first post, I also used swap space (gigabytes worth, I think) and then it crashed.

I don't have access to the pi zero 2w at the moment but I can test on a computer with any amount of memory to get the same result (probably), provided I supply enough images. If you want, I can compile and test it on a riscv board with 512 MB of ram, the next lowest amount I can test with now is 4 GB (arm64 but it can also run armhf).

**Comment by leejuyuu** [comment]
> you can easily monitor memory use and see if it goes above a certain value that you set

As I said, the current implementation does not respect the configured cache limit strictly and the `maxImagesCached` needs to be considered as well. 

I understand this back-and-forth is frustrating. However, I do not have an arm-based device, therefore I need your help doing these tests. Please try altering the `maxImagesCached` and/or setting cache limit to 0.

I currently do not see anywhere leaking. The image loader allocates most of the memory, but on my system this is (somewhat) bounded and older cache does get freed when browsing. My requests were meant to determine whether the OOM issue you are facing is due to how nomacs caches images, which could be controlled by configuration. If it isn't, then we'll know something more serious or arm platform-specific is going on, and probably need to get one of those devices for myself :))

**Comment by lukaszsobala** [comment]
@leejuyuu ok, now I understand. You can't reproduce this (but I'm confused as @scrubbbbs did seem to reproduce it).

I might have omitted some context: the folder where the images were contained only symlinks to the images, not the actual images. Does this reproduce the leak at your side?

**Comment by leejuyuu** [comment]
Might actually be related to the thumbnail. They get created even without opening the panel. Because each thumbnail is small, the number of files in the directory is important. Working on a fix

Heap profile
![heap](https://github.com/user-attachments/assets/2145858f-7f41-475d-a576-575e140a6a77)

**Comment by lukaszsobala** [comment]
OK then, this makes sense. Let me know when I can pull and test, and also if it's worth for you that I test on RISC-V. I can do this natively (on Milk-V Duo S) or in an ubuntu 24.10 RISC-V virtual machine.

**Comment by leejuyuu** [comment]
Sure, I'll let you know when a PR is ready. The testing could be deferred until then. 

> Please also remember that I don't know much about programming so any jargon might not be understandable for me.

Sorry about that. If anything is unclear, please feel free to ask so we can clarify.

**Comment by leejuyuu** [comment]
@lukaszsobala, you can test https://github.com/nomacs/nomacs/pull/1262. It's still in progress, but the memory growth should be fixed now (if you do not open the thumbnail panel).

**Comment by lukaszsobala** [comment]
Thanks @leejuyuu, sorry for the question but to check it do I need to do `git checkout #1262` and build?

**Comment by leejuyuu** [comment]
If using git, I would do

```bash
git remote add leejuyuu https://github.com/leejuyuu/nomacs.git
git fetch leejuyuu
git checkout leejuyuu/thumbs
```

**Comment by leejuyuu** [comment]
@lukaszsobala, have you got a chance to see if it solves the memory issue for you?

**Comment by lukaszsobala** [comment]
@leejuyuu I haven't checked yet, sorry, I am quite busy until the end of February. Thanks for the guidance so far

**Comment by leejuyuu** [comment]
This should be fixed by #1262. Feel free to reopen if there are further issues.
